src/System.Memory/src/System/Buffers/ReadOnlyBuffer.cs

```diff
@@ -0,0 +1,417 @@
+
+        internal const int OwnedMemoryStartMask = 1 << 31;
+        internal const int OwnedMemoryEndMask = 0;
+
+        internal readonly SequencePosition BufferStart;
+        internal readonly SequencePosition BufferEnd;
+
+        public static readonly ReadOnlyBuffer<T> Empty = new ReadOnlyBuffer<T>(new T[0]);
+
+        /// <summary>
+        /// Length of the <see cref="ReadOnlyBuffer{T}"/> in bytes.
+        /// </summary>
+        public long Length => GetLength(BufferStart, BufferEnd);
+
+        /// <summary>
+        /// Determines if the <see cref="ReadOnlyBuffer{T}"/> is empty.
+        /// </summary>
+        public bool IsEmpty => Length == 0;
+
+        /// <summary>
+        /// Determins if the <see cref="ReadOnlyBuffer{T}"/> is a single <see cref="Memory{Byte}"/>.
+        /// </summary>
+        public bool IsSingleSegment => BufferStart.Segment == BufferEnd.Segment;
+
+        public ReadOnlyMemory<T> First
+        {
+            get
+            {
+                TryGetBuffer(BufferStart, BufferEnd, out var first, out _);
+                return first;
+            }
+        }
+
+        /// <summary>
+        /// A position to the start of the <see cref="ReadOnlyBuffer{T}"/>.
+        /// </summary>
+        public SequencePosition Start => BufferStart;
+
+        /// <summary>
+        /// A position to the end of the <see cref="ReadOnlyBuffer{T}"/>
+        /// </summary>
+        public SequencePosition End => BufferEnd;
+
+        private ReadOnlyBuffer(object startSegment, int startIndex, object endSegment, int endIndex)
+        {
+            Debug.Assert(startSegment != null);
+            Debug.Assert(endSegment != null);
+
+            BufferStart = new SequencePosition(startSegment, startIndex);
+            BufferEnd = new SequencePosition(endSegment, endIndex);
+        }
+
+        public ReadOnlyBuffer(IMemoryList<T> startSegment, int offset, IMemoryList<T> endSegment, int endIndex)
+        {
+            Debug.Assert(startSegment != null);
+            Debug.Assert(endSegment != null);
+            Debug.Assert(startSegment.Memory.Length >= offset);
+            Debug.Assert(endSegment.Memory.Length >= endIndex);
+
+            BufferStart = new SequencePosition(startSegment, offset | MemoryListStartMask);
+            BufferEnd = new SequencePosition(endSegment, endIndex | MemoryListEndMask);
+        }
+
+        public ReadOnlyBuffer(T[] array) : this(array, 0, array.Length)
+        {
+        }
+
+        public ReadOnlyBuffer(T[] array, int offset, int length)
+        {
+            if (array == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.array);
+            }
+
+            BufferStart = new SequencePosition(array, offset | ArrayStartMask);
+            BufferEnd = new SequencePosition(array, offset + length | ArrayEndMask);
+        }
+
+        public ReadOnlyBuffer(Memory<T> data)
+        {
```

Comment:
We should add doc comments explaining lifetime guarantees of the memory, i.e. the caller is supposed to ensure that the memory lives longer than this buffer.